R/cali_ensemble.R

```diff
@@ -418,40 +421,94 @@ cali_ensemble <- function(config_file, num = NULL, param_file = NULL, cmethod =
 
   names(met_l) <- model
 
-##------------------------- parallel LCH calibration -----------------------------------------------
+##------------------------- parallel LHC calibration -----------------------------------------------
 
   if(parallel){
-    ncores <- detectCores() - 1
-    clust <- makeCluster(ncores)
-    clusterExport(clust, varlist = list("pars_lhc", "pars_l", "model", "config_file", "met_l",
+    
+    if (is.null(ncores)) {
+      ncores <- parallel::detectCores() - 1
+    }
+    
+    cl <- parallel::makeCluster(ncores)
+    on.exit(parallel::stopCluster(cl))
+    parallel::clusterExport(cl, varlist = list("pars_lhc", "pars_l", "model", "config_file", "met_l",
                                         "folder", "out_f", "cnfg_l", "obs_deps",
                                         "obs_out", "out_hour", "qualfun",
                                         "outf_n"),
                   envir = environment())
-    clusterEvalQ(clust, library(LakeEnsemblR))
-    if(cmethod == "LHC") {
-      message("\nStarted parallel LHC\n")
+    parallel::clusterExport(cl = cl, 
+                            unclass(lsf.str(envir = asNamespace("LakeEnsemblR"), 
+                                            all = T)),
+                            envir = as.environment(asNamespace("LakeEnsemblR"))
+    )
+    
+    if (cmethod == "LHC") {
+      
       model_out <- setNames(
-        parLapply(clust, model, function(m) LHC_model(pars = pars_lhc[[m]],
-                                                      type = pars_l[[m]]$type,
-                                                      model = m, var = "temp",
-                                                      config_file = config_file,
-                                                      met = met_l[[m]], folder = folder,
-                                                      out_f = out_f, config_f = cnfg_l[[m]],
-                                                      obs_deps = obs_deps, obs_out = obs_out,
-                                                      out_hour = out_hour, qualfun = qualfun,
-                                                      nout_fun = nout_fun, outf_n = outf_n
-        )),
-        model
-      )
-      message("\nFinished parallel LHC\n")
+        lapply(model, \(m) {
```

Comment:
I'm wondering if it would be better to either create a new function or at least create a new `cmethod` in order to keep the old LHC method and prevent destroying workflows of other people...